lib/kino/utils/table.ex

```diff
@@ -0,0 +1,121 @@
+defmodule Kino.Utils.Table do
+  @moduledoc false
+
+  # Common functions for handling various Elixir
+  # terms as table records.
+
+  @doc """
+  Computes table columns that accomodate for all the given records.
+  """
+  def columns_for_records(records) do
+    case Enum.at(records, 0) do
+      nil ->
+        []
+
+      first_record ->
+        first_record_columns = columns_for_record(first_record)
+
+        all_columns =
+          records
+          |> Enum.reduce(MapSet.new(), fn record, columns ->
+            record
+            |> columns_for_record()
+            |> MapSet.new()
+            |> MapSet.union(columns)
+          end)
+          |> MapSet.to_list()
+          |> Enum.sort_by(& &1.key)
+
+        # If all records have the same structure, keep the order,
+        # otherwise return the sorted accumulated columns
+        if length(first_record_columns) == length(all_columns) do
+          first_record_columns
+        else
+          all_columns
+        end
+    end
+  end
+
+  defp columns_for_record(record) when is_tuple(record) do
+    record
+    |> Tuple.to_list()
+    |> Enum.with_index()
+    |> Enum.map(&elem(&1, 1))
+    |> keys_to_columns()
+  end
+
+  defp columns_for_record(record) when is_map(record) do
+    if is_struct(record) and ecto_schema?(record.__struct__) do
+      record.__struct__.__schema__(:fields)
+    else
+      record |> Map.keys() |> Enum.sort()
+    end
+    |> keys_to_columns()
+  end
+
+  defp columns_for_record(record) when is_list(record) do
+    record
+    |> Keyword.keys()
+    |> keys_to_columns()
+  end
+
+  defp columns_for_record(_record) do
+    # If the record is neither of the expected enumerables,
+    # we treat it as a single column value
+    keys_to_columns([:item])
+  end
+
+  @doc """
+  Converts keys to column specifications.
+  """
+  def keys_to_columns(keys) do
+    Enum.map(keys, fn key -> %{key: key, label: inspect(key)} end)
+  end
+
+  @doc """
+  Looks up record field value by key.
+  """
+  def get_field(record, key)
+
+  def get_field(record, key) when is_tuple(record) do
+    if key < tuple_size(record) do
+      elem(record, key)
+    else
+      nil
+    end
+  end
+
+  def get_field(record, key) when is_list(record) do
+    record[key]
+  end
+
+  def get_field(record, key) when is_map(record) do
+    Map.get(record, key)
+  end
+
+  def get_field(record, :item) do
+    record
+  end
+
+  @doc """
+  Converts a record to row specification given a list
+  of desired keys.
+  """
+  def record_to_row(record, keys) do
+    fields =
+      Map.new(keys, fn key ->
+        value = get_field(record, key)
+        {key, inspect(value)}
+      end)
+
+    # Note: id is opaque to the client, and we don't need it for now
+    %{id: nil, fields: fields}
+  end
+
+  @doc """
+  Checks if the given term is an Ecto.Schema.
+  """
+  def ecto_schema?(queryable) do
```

Comment:
Is this reused anywhere? If not, maybe we should move to the Ecto module?

Reply:
It's used in this module in `columns_for_record`.

Reply:
Hrm, not sure if it is worth having it there. It is like the handling of Ecto queries is now everywhere instead of isolated in a single place. Consider splitting that apart or having Ecto not call it?

Also remember that Ecto’s select can return a tuple, so we need to take that into account too.

Reply:
This was kinda intentional, because if someone passes a list of Ecto structs into `DataTable`, we may as well use the fields information it provides, right?

> Also remember that Ecto’s select can return a tuple, so we need to take that into account too.

Yeah, it can be a map or a single value too, that's the reason why I moved much stuff to `Table.Utils`, because it's pretty much the same for `DataTable` and `ETS`.

Reply:
> This was kinda intentional, because if someone passes a list of Ecto structs into DataTable, we may as well use the fields information it provides, right?

i would just treat them as structs since the fields are in the struct too? What benefits would be added by looking into Ecto?

Reply:
The benefits are that the columns are ordered as per the schema definition, which should feel more natural. Also this way we don't show virtual fields such as associations.

Reply:
Hrm, I see. Sounds good to me then!